### PR TITLE
[IE CLDNN] Add asymmetric quantization support to fsv16 imad general convolution kernel

### PR DESCRIPTION
**Pull request by jhajducz**
This patch is meant to:
- add asymmetric quantization support (for both activations and weights) with and without using compensation term to convolution_gpu_b_fs_zyx_fsv16_imad kernel and add/modify corresponding unit tests

JIRA: CVS-25121

### CODE REVIEW

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_zyx_fsv16_imad.cpp [hunk 1]
```diff
@@ -320,6 +320,9 @@ ParamsKey Convolution_kernel_b_fs_zyx_fsv16_imad::GetSupportedKey() const {
     k.EnableBatching();
     k.EnableGroupedConvolution();
     k.EnableQuantization(QuantizationType::SYMMETRIC);
+    k.EnableQuantization(QuantizationType::ASYMMETRIC_DATA);
+    k.EnableQuantization(QuantizationType::ASYMMETRIC_WEIGHTS);
+    k.EnableQuantization(QuantizationType::ASYMMETRIC_DATA_AND_WEIGHTS);
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_zyx_fsv16_imad.cpp:325]: Don't you need to adjust layout optimizer logic to enable this kernel with asymmetric quantization?

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl [hunk 2]
```diff
@@ -40,6 +55,12 @@ KERNEL(convolution_gpu_b_fs_zyx_fsv16_imad)(
     const __global FILTER_TYPE *weights,
 #if BIAS_TERM
     const __global BIAS_TYPE *biases,
+#endif
+    #if ASYMMETRIC_WEIGHTS_QUANTIZATION
+    const __global WEIGHTS_ZERO_POINTS_TYPE *weights_zp,
+#endif
+#if ASYMMETRIC_DATA_QUANTIZATION
+    const __global ACTIVATIONS_ZERO_POINTS_TYPE *activations_zp,
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl:63]: one argument (compensation) is missing. How does it work? I suppose if this kernel is somehow selected, then you'll have setArg error

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl [hunk 3]
```diff
@@ -40,6 +55,12 @@ KERNEL(convolution_gpu_b_fs_zyx_fsv16_imad)(
     const __global FILTER_TYPE *weights,
 #if BIAS_TERM
     const __global BIAS_TYPE *biases,
+#endif
+    #if ASYMMETRIC_WEIGHTS_QUANTIZATION
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl:59]: nit: alignment

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_zyx_fsv16_imad.cpp [hunk 4]
```diff
@@ -422,11 +425,25 @@ bool Convolution_kernel_b_fs_zyx_fsv16_imad::Validate(const Params& params, cons
...
+    convolution_params& conv_params = *static_cast<convolution_params*>(kd.params.get());
 
-    if (newParams.split != 1)
+    if (conv_params.split != 1)
         return false;
 
+    if (conv_params.quantization == QuantizationType::ASYMMETRIC_DATA_AND_WEIGHTS) {
+        if (conv_params.activations_zero_points.empty() || conv_params.weights_zero_points.empty())
+            return false;
+    } else if (conv_params.quantization == QuantizationType::ASYMMETRIC_DATA) {
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_zyx_fsv16_imad.cpp:436]: Here you should also check that `!conv_params.compensation.empty()` for both ASYMMETRIC_DATA and ASYMMETRIC_DATA_AND_WEIGHTS quantization types

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl [hunk 5]
```diff
@@ -92,8 +113,64 @@ KERNEL(convolution_gpu_b_fs_zyx_fsv16_imad)(
...
 
+#if ASYMMETRIC_DATA_QUANTIZATION
+    uint data_zp_idx = g * FILTER_IFM_NUM + in_f_start;
+    uint4 data_zp_val;
+#endif
+#if ASYMMETRIC_WEIGHTS_QUANTIZATION
+    uint4 weights_zp_val[OFM_BLOCKS_PER_SIMD];
+    __attribute__((opencl_unroll_hint))
+    for (uint ofb = 0; ofb < OFM_BLOCKS_PER_SIMD; ++ofb) {
+        weights_zp_val[ofb] = as_uint4((FILTER_TYPE_16)weights_zp[out_f + ofb * FSV]);
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl:124]: Have you checked how big is overhead for loading per-channel zero-points? Right now we always create per-channel parameters, but actually zero point can be per-tensor, which might be much better from the performance perspective.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl [hunk 6]
```diff
@@ -146,11 +219,11 @@ KERNEL(convolution_gpu_b_fs_zyx_fsv16_imad)(
...
                                         } else {
-                                            input_int8_arr[v] = conv_input[input_idx + tmp * FSV + v + 
-                                                                           ((INPUT0_SIZE_X + 2*PADDING_SIZE_X) * 
-                                                                            (INPUT0_SIZE_Y + 2*PADDING_SIZE_Y) * 
-                                                                            (INPUT0_SIZE_Z + 2*PADDING_SIZE_Z) - 1) * 
-                                                                           FSV];
+                                            const uint addr = input_idx + tmp * FSV + v +
+                                                        ((INPUT0_SIZE_X + INPUT0_PAD_BEFORE_SIZE_X + INPUT0_PAD_AFTER_SIZE_X) *
+                                                         (INPUT0_SIZE_Y + INPUT0_PAD_BEFORE_SIZE_Y + INPUT0_PAD_AFTER_SIZE_Y) *
+                                                         (INPUT0_SIZE_Z + INPUT0_PAD_BEFORE_SIZE_Z + INPUT0_PAD_AFTER_SIZE_Z) - 1) * FSV;
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl:225]: Could you extract accuracy fix into separate PR?

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl [hunk 7]
```diff
@@ -514,3 +644,7 @@ KERNEL(convolution_gpu_b_fs_zyx_fsv16_imad)(
 #undef SIMD
 #undef FSV
 #undef OFM_VALUES_PER_WI
+
+#if INPUT0_PAD_BEFORE_SIZE_X != 0 || INPUT0_PAD_BEFORE_SIZE_Y != 0 || INPUT0_PAD_BEFORE_SIZE_Z != 0
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl:648]: `#ifdef NON_ZERO_INPUT0_PAD_BEFORE`

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl [hunk 8]
```diff
@@ -503,9 +620,22 @@ KERNEL(convolution_gpu_b_fs_zyx_fsv16_imad)(
...
 #undef AS_TYPE_N_
+
+#undef INPUT0_TYPE_4
+#undef AS_INPUT0_TYPE_4
+
+#if ASYMMETRIC_DATA_QUANTIZATION && ASYMMETRIC_WEIGHTS_QUANTIZATION
+#undef ACCUMULATOR_TYPE_4
+#endif
+
+#if ASYMMETRIC_WEIGHTS_QUANTIZATION
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl:635]: `#ifdef`

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl [hunk 9]
```diff
@@ -503,9 +620,22 @@ KERNEL(convolution_gpu_b_fs_zyx_fsv16_imad)(
...
-#undef AS_INPUT0_TYPE_4
+#undef TYPE_N_
+#undef TYPE_N
 #undef AS_TYPE_N
 #undef AS_TYPE_N_
+
+#undef INPUT0_TYPE_4
+#undef AS_INPUT0_TYPE_4
+
+#if ASYMMETRIC_DATA_QUANTIZATION && ASYMMETRIC_WEIGHTS_QUANTIZATION
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl:631]: `#ifdef`

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl (hunk 6 above)
Comment (jhajducz) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl:225]: Done in #2876

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_zyx_fsv16_imad.cpp (hunk 1 above)
Comment (jhajducz) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_zyx_fsv16_imad.cpp:325]: Fixed

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl (hunk 2 above)
Comment (jhajducz) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl:80]: Fixed

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl (hunk 3 above)
Comment (jhajducz) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl]: Fixed

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_zyx_fsv16_imad.cpp (hunk 4 above)
Comment (jhajducz) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_zyx_fsv16_imad.cpp]: Fixed

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl (hunk 5 above)
Comment (jhajducz) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl:145]: I haven't run such measurements, I assumes that we need to have per-channel parameters.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl (hunk 7 above)
Comment (jhajducz) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl]: Fixed

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl (hunk 8 above)
Comment (jhajducz) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl]: Fixed

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl (hunk 9 above)
Comment (jhajducz) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_zyx_fsv16_imad.cl]: Fixed

**Review by vladimir-paramuzov — APPROVED** [review]
